File: modules/core/src/main/scala/RequestContext.scala
/*
 * Copyright 2021 Hossein Naderi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edomata.core

import cats.data.NonEmptyChain
import cats.implicits.*

import java.time.Instant

/** Representation of a standard input for an [[Edomaton]]
  *
  * @tparam C
  *   Command type
  * @tparam S
  *   State model type
  */
final case class RequestContext[+C, +S](
    command: CommandMessage[C],
    state: S
)

/** Representation of a standard command message
  *
  * @tparam C
  *   Command payload which is your command model
  */
final case class CommandMessage[+C](
    id: String,
    time: Instant,
    address: String,
    payload: C,
    metadata: MessageMetadata
)

object CommandMessage {

  /** Constructs a command message a the root of a chain of messages */
  def apply[C](
      id: String,
      time: Instant,
      address: String,
      payload: C
  ): CommandMessage[C] = CommandMessage(
    id = id,
    time = time,
    address = address,
    payload,
    MessageMetadata(id)
  )

  extension [C, M](cmd: CommandMessage[C]) {

    /** Builds a request context from this [[CommandMessage]] */
    def buildContext[S, R](
        state: S
    ): RequestContext[C, S] =
      RequestContext(
        command = cmd,
        state = state
      )

    /** Derives a new metadata for the next message in chain that is linked to
      * this one
      */
    def deriveMeta: MessageMetadata = cmd.metadata.copy(causation = cmd.id.some)
  }
}

/** Representation of a standard message metadata */
final case class MessageMetadata(
    correlation: Option[String],
    causation: Option[String]
)
object MessageMetadata {

  /** Consturcts a root chain metadata */
  def apply(id: String): MessageMetadata = MessageMetadata(id.some, id.some)

  /** Consturcts a metadata */
  def apply(correlation: String, causation: String): MessageMetadata =
    MessageMetadata(correlation = correlation.some, causation = causation.some)
}

File: rust/crates/edomata-core/src/request_context.rs
//! Command messages and the request context handed to domain programs.

use chrono::{DateTime, Utc};

/// The input of an [`Edomaton`](crate::Edomaton): the command being handled
/// together with the current state of the aggregate.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RequestContext<C, S> {
    /// The command message.
    pub command: CommandMessage<C>,
    /// Current state of the aggregate.
    pub state: S,
}

impl<C, S> RequestContext<C, S> {
    /// Builds a request context.
    pub fn new(command: CommandMessage<C>, state: S) -> Self {
        Self { command, state }
    }
}

/// A command message sent to an aggregate.
#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct CommandMessage<C> {
    /// Unique identifier of this message; used for idempotency.
    pub id: String,
    /// When the command was issued.
    pub time: DateTime<Utc>,
    /// Address of the target aggregate (its stream identifier).
    pub address: String,
    /// The command payload, your command model.
    pub payload: C,
    /// Correlation and causation metadata.
    pub metadata: MessageMetadata,
}

impl<C> CommandMessage<C> {
    /// Constructs a command message at the root of a chain of messages.
    pub fn new(
        id: impl Into<String>,
        time: DateTime<Utc>,
        address: impl Into<String>,
        payload: C,
    ) -> Self {
        let id = id.into();
        let metadata = MessageMetadata::root(id.clone());
        Self {
            id,
            time,
            address: address.into(),
            payload,
            metadata,
        }
    }

    /// Constructs a command message with explicit metadata.
    pub fn with_metadata(
        id: impl Into<String>,
        time: DateTime<Utc>,
        address: impl Into<String>,
        payload: C,
        metadata: MessageMetadata,
    ) -> Self {
        Self {
            id: id.into(),
            time,
            address: address.into(),
            payload,
            metadata,
        }
    }

    /// Builds a request context from this message.
    pub fn build_context<S>(self, state: S) -> RequestContext<C, S> {
        RequestContext {
            command: self,
            state,
        }
    }

    /// Derives the metadata of the next message in the chain that is caused
    /// by this one.
    pub fn derive_meta(&self) -> MessageMetadata {
        MessageMetadata {
            correlation: self.metadata.correlation.clone(),
            causation: Some(self.id.clone()),
        }
    }

    /// Changes the payload.
    pub fn map<D, F: FnOnce(C) -> D>(self, f: F) -> CommandMessage<D> {
        CommandMessage {
            id: self.id,
            time: self.time,
            address: self.address,
            payload: f(self.payload),
            metadata: self.metadata,
        }
    }
}

/// Correlation and causation identifiers of a message.
#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct MessageMetadata {
    /// Identifier of the whole chain of messages.
    pub correlation: Option<String>,
    /// Identifier of the message that directly caused this one.
    pub causation: Option<String>,
}

impl MessageMetadata {
    /// Constructs the metadata of a root message: correlation and causation
    /// are both the message id.
    pub fn root(id: impl Into<String>) -> Self {
        let id = id.into();
        Self {
            correlation: Some(id.clone()),
            causation: Some(id),
        }
    }

    /// Constructs metadata with explicit identifiers.
    pub fn new(correlation: impl Into<String>, causation: impl Into<String>) -> Self {
        Self {
            correlation: Some(correlation.into()),
            causation: Some(causation.into()),
        }
    }

    /// Metadata without any identifier.
    pub fn empty() -> Self {
        Self::default()
    }
}
